xwiki-commons-core/xwiki-commons-environment/xwiki-commons-environment-servlet/src/main/java/org/xwiki/environment/internal/ServletEnvironment.java

```diff
@@ -54,6 +62,41 @@ public class ServletEnvironment extends AbstractEnvironment
      */
     private javax.servlet.ServletContext javaxServletContext;
 
+    @Inject
+    private ComponentManager componentManager;
+
+    @Inject
+    private CacheControl cacheControl;
+
+    private Cache<Optional<URL>> resourceURLCache;
+
+    private final AtomicBoolean cacheInitializing = new AtomicBoolean();
+
+    private Cache<Optional<URL>> getResourceURLCache()
+    {
+        // Don't block on cache initialization to avoid loops.
+        if (this.resourceURLCache == null && this.cacheInitializing.compareAndSet(false, true)) {
+            try {
+                // Compare again inside the "lock" to avoid race conditions. Only after seeing the false value we can
+                // be sure that we also see the resourceURLCache variable write from the thread that wrote it -
+                // writing atomics with "set" guarantees sequential consistency.
+                if (this.resourceURLCache == null) {
+                    // The cache manager can't be injected directly as it depends on this component, so it is
+                    // important to only request it once this component has been initialized.
+                    CacheManager cacheManager = this.componentManager.getInstance(CacheManager.class);
+                    this.resourceURLCache = cacheManager.createNewCache(
+                        new LRUCacheConfiguration("environment.servlet.resourceURLCache", 1000));
```

Comment:
Isn't 1000 a bit low value by default?

Reply:
At least it feels like we could go up without it being noticeable much in terms of memory.

Reply:
10k sounds good

Reply:
:+1: was also thinking of 10K